### PR TITLE
Update to python 3.9

### PR DESCRIPTION
**Pull request by MrCreosote**
Needed to update a few libraries for compatibility

Also minor simplifications to a test fixture

### CODE REVIEW

File: tests/test_autodetect.py [hunk 1]
```diff
@@ -8,8 +8,8 @@
 from tests.test_utils import bootstrap_config
 
 
-@pytest.fixture(autouse=True)
-def run_before_and_after_tests(tmpdir):
+@pytest.fixture(autouse=True, scope="module")
+def run_before_tests():
```
Comment (MrCreosote) [inline tests/test_autodetect.py:12]: Only needs to run at the beginning of the tests AFAICT

**Comment by MrCreosote** [comment]
LGTM is complaining about files this PR doesn't touch

**Comment by bio-boris** [comment]
Might as well upgrade it to 3.9, it works for me for `FROM python:3.9`

**Comment by MrCreosote** [comment]
I did 3.8 because 3.9 isn't in apt for my distribution yet, and the features I want are in 3.7. If I need 3.9 I'll go through the hassle of installing a ppa then

**Comment by bio-boris** [comment]
I'm in favor of taking it all the way to 3.9 if we are upgrading python.
It shouldn't take more than a few minutes to install miniconda for 3.9, setup a virtualenv and install requirements.txt, or to just use the python in the docker container. 
I don't think we have any policy on upgrading services based on using python versions with the latest  EOL vs developer convenience, but  otherwise the PR looks fine and deployed to CI

**Comment by MrCreosote** [comment]
I wouldn't touch *conda if you served it to me on a skewer with hollandaise sauce. Several times bitten, several times shy

**Comment by MrCreosote** [comment]
Turns out I need features in 3.9 as well, so you get your wish

**Pull request merged**